File: 0207-course-schedule/0207-course-schedule.py
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        prereqCnt = {k: 0 for k in range(numCourses)}
        graph = {k: [] for k in range(numCourses)}

        for (p, s) in prerequisites:
            graph[p].append(s)
            prereqCnt[s] += 1

        while True:
            start: Optional[int] = None
            for (i, c) in prereqCnt.items():
                if c == 0:
                    start = i
                    break

            if start is None:
                break

            for s in graph[start]:
                if s in prereqCnt:
                    prereqCnt[s] -= 1
            del graph[start]
            del prereqCnt[start]

        return len(prereqCnt) == 0

File: 0207-course-schedule/0207-course-schedule.kt
class Solution {
    fun canFinish(numCourses: Int, prerequisites: Array<IntArray>): Boolean {
        val edges = mutableMapOf<Int, MutableSet<Int>>().apply {
            (0..<numCourses).forEach {
                this[it] = mutableSetOf()
            }
        }
        val ins = IntArray(numCourses)

        prerequisites.forEach { (u, v) ->
            if (v !in edges[u]!!) {
                edges[u]!!.add(v)
                ins[v]++
            }
        }

        val next = ArrayDeque<Int>().apply {
            addAll(ins.withIndex().filter { (_, c) -> c == 0 }.map { (i, _) -> i })
        }
        while (next.isNotEmpty()) {
            val p = next.removeFirst()
            val neighbors = edges[p]!!
            edges[p]!!.forEach {
                ins[it]--
                if (ins[it] == 0) {
                    next.add(it)
                }
            }
        }

        return ins.all { it == 0 }
    }
}
